pkg/controller/bluegreendeployment/controller.go

```diff
@@ -0,0 +1,195 @@
+package bluegreendeployment
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+	"time"
+
+	"github.com/kubernetes-sigs/kubebuilder/pkg/controller"
+	"github.com/kubernetes-sigs/kubebuilder/pkg/controller/types"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	controllerv1alpha1 "k8s.io/bgd/pkg/apis/controller/v1alpha1"
+	controllerv1alpha1client "k8s.io/bgd/pkg/client/clientset_generated/clientset/typed/controller/v1alpha1"
+	"k8s.io/bgd/pkg/inject/args"
+	bgdutil "k8s.io/bgd/pkg/util"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	BlueColor               = "blue"
+	GreenColor              = "green"
+	BlueGreenDeploymentKind = "BlueGreenDeployment"
+	BlueRSName              = BlueColor + "-rs"
+	GreenRSName             = GreenColor + "-rs"
+	ServiceName             = "bgd-svc"
+
+	LabelColorKey = "color"
+
+	PollInterval = 100 * time.Millisecond
+	PollTimeout  = 30 * time.Second
+)
+
+// Reconcile reconciles BlueGreenDeployment object to desired state.
+func (c *BlueGreenDeploymentController) Reconcile(k types.ReconcileKey) error {
+	b, err := c.blueGreenDeploymentClient.BlueGreenDeployments(k.Namespace).Get(k.Name, metav1.GetOptions{})
+	if err != nil || b == nil {
+		return err
+	}
+
+	if b, err = defaultFields(c.blueGreenDeploymentClient, b); err != nil {
+		return err
+	}
+
+	if err := c.sync(b); err != nil {
+		log.Fatalln(err)
+		return err
+	}
+
+	return nil
+}
+
+// +controller:group=controller,version=v1alpha1,kind=BlueGreenDeployment,resource=bluegreendeployments
+// +rbac:groups=controller,resources=bluegreendeployments,verbs=get;list;watch;create;update;patch;delete
```

Comment:
you should get the rbac for the resource you are defining for free, but you will need to enable the rbac for the things you are creating / watching - e.g. replicasets

Reply:
i see - it means ReplicaSets, pods, and service here. :)